plugin/src/main/java/org/opensearch/ml/action/connector/UpdateConnectorTransportAction.java

```diff
@@ -86,31 +100,57 @@ public UpdateConnectorTransportAction(
-                    updateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
-                    updateRequest.doc(connector.toXContent(XContentBuilder.builder(XContentType.JSON.xContent()), ToXContent.EMPTY_PARAMS));
-                    updateUndeployedConnector(connectorId, updateRequest, listener, context);
-                } else {
-                    listener
-                        .onFailure(
-                            new IllegalArgumentException("You don't have permission to update the connector, connector id: " + connectorId)
-                        );
-                }
-            }, exception -> {
-                log.error("Permission denied: Unable to update the connector with ID {}. Details: {}", connectorId, exception);
-                listener.onFailure(exception);
-            }));
+            connectorAccessControlHelper
+                .getConnector(sdkClient, client, context, getDataObjectRequest, connectorId, ActionListener.wrap(connector -> {
+                    // context is already restored here
+                    if (TenantAwareHelper.validateTenantResource(mlFeatureEnabledSetting, tenantId, connector.getTenantId(), listener)) {
+                        boolean hasPermission = connectorAccessControlHelper.validateConnectorAccess(client, connector);
+                        if (hasPermission) {
+                            connector.update(mlUpdateConnectorAction.getUpdateContent(), mlEngine::encrypt);
+                            connector.validateConnectorURL(trustedConnectorEndpointsRegex);
+                            connector.setLastUpdateTime(Instant.now());
+                            UpdateDataObjectRequest updateDataObjectRequest = UpdateDataObjectRequest
+                                .builder()
+                                .index(ML_CONNECTOR_INDEX)
+                                .id(connectorId)
+                                .tenantId(tenantId)
+                                .dataObject(connector)
+                                .build();
+                            try (ThreadContext.StoredContext innerContext = client.threadPool().getThreadContext().stashContext()) {
+                                updateUndeployedConnector(
+                                    connectorId,
+                                    updateDataObjectRequest,
+                                    ActionListener.runBefore(listener, innerContext::restore)
+                                );
+                            }
+                        } else {
+                            listener
+                                .onFailure(
+                                    new IllegalArgumentException(
+                                        "You don't have permission to update the connector, connector id: " + connectorId
+                                    )
+                                );
+                        }
```

Comment:
fyi This might not work
```suggestion
    if (!hasPermission){
        listener.onFailure(
            new IllegalArgumentException(
                "You don't have permission to update the connector, connector id: " + connectorId
            )
        );
        return; // not sure if we can early return here
    }

    connector.update(mlUpdateConnectorAction.getUpdateContent(), mlEngine::encrypt);
    connector.validateConnectorURL(trustedConnectorEndpointsRegex);
    connector.setLastUpdateTime(Instant.now());

    UpdateDataObjectRequest updateDataObjectRequest = UpdateDataObjectRequest
        .builder()
        .index(ML_CONNECTOR_INDEX)
        .id(connectorId)
        .tenantId(tenantId)
        .dataObject(connector)
        .build();
        
    try (ThreadContext.StoredContext innerContext = client.threadPool().getThreadContext().stashContext()) {
        updateUndeployedConnector(
            connectorId,
            updateDataObjectRequest,
            ActionListener.runBefore(listener, innerContext::restore)
        );
    }
        
```

Reply:
IMO this method is already readable enough.